content/electrical/dc.md

```diff
@@ -0,0 +1,543 @@
+
+
+```{code-cell} ipython3
+Pmax = np.max(Pload) # Finding the maximum value of Pload
+print(f"Pmax = {Pmax}")
+
+Pmax_idx = np.argmax(Pload) # Finding the index in which Pload is maximum
+print(f"Pmax_idx = {Pmax_idx}")
+
+Rload[Pmax_idx]
+print(f"Rload[Pmax_idx] = {Rload[Pmax_idx]}")
+```
+
+```{admonition} Challenge
+ Why is the value of `Rload` not exactly 5? How would you
+increase the accuracy?
+```
+
+## Part 3: Solving circuits using mesh analysis
+
+### Part 3a: 2x2 Mesh Matrix
+
+```{figure} imgs/circuit4.png
+---
+width: 400px
+height: 200px
+---
+Figure 4: Loop Circuit
+```
+
+Doing a mesh analysis on *Figure 4* yields *two linear equations*:
+
+
+```{math}
+I1(R1+R2) + I2(-R2) = V1
+
+I1(-R2) + I2(R2+R3) = -V2
+```
+
+These two equations can be written in *matrix format* and inputted into NumPy:
+
+```{math}
+\begin{bmatrix} R1+R2 & -R2 \\ -R2 & R2+R3 \end{bmatrix} 
+* 
+\begin{bmatrix} I1 \\ I2  \end{bmatrix}
+=
+\begin{bmatrix} V1 \\ -V2  \end{bmatrix}
+```
+
+```{admonition} Note
+Check out Wikipedia's articles on 
+[mesh analysis](https://www.electronics-tutorials.ws/dccircuits/dcp_5.html),
+[linear equations](https://en.wikipedia.org/wiki/System_of_linear_equations), 
+and [matrices](https://en.wikipedia.org/wiki/Matrix_(mathematics))
+if you would like to revise the concepts presented here
+```
+
+```{code-cell} ipython3
+R1 = 10
+R2 = 5
+R3 = 5
+V1 = 5
+V2 = -15
+
+a11 = R1+R2
+a12 = -R2
+a21 = -R2
+a22 = R2+R3
+
+b1 = V1
+b2 = -V2
+
+A = np.array([[a11, a12],
+              [a21, a22]])
+
+b = np.array([[b1], [b2]])
+
+print(A)
+print(b)
+```
+
+If solving the matrix using pencil and paper, you would first invert $A$ and
+then multiply it by $b$.
+
+```{math}
+\begin{bmatrix} I1 \\ I2  \end{bmatrix} 
+= 
+\begin{bmatrix} R1+R2 & -R2 \\ -R2 & R2+R3 \end{bmatrix}^{-1} 
+* 
+\begin{bmatrix} V1 \\ -V2  \end{bmatrix}
+```
+
+```{code-cell} ipython3
+Ainv = np.linalg.inv(A)
+print(Ainv)
+
+x = Ainv @ b # The @ operator multiplies two matrices
+print(x)
+```
+
+```{figure} imgs/pikachu_angry.jpg
+---
+width: 500px
+---
+
+```
+
+**NEVER** invert your matrices if you are programming on a computer (except if
+you actually need the inverted matrix itself, which is rare). The process is
+very sensitive to scaling, propagates numerical errors and will make **Pikachu**
+angry! Instead use the `np.linalg.solve` function.
+
+```{admonition} Note 1
+If you want to learn more about the complications arising from matrix inversion,
+check out this
+(article)[https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/]
+talking about it's harms. And check out this
+(paper)[http://www.netlib.org/lapack/lawnspdf/lawn27.pdf] if you want a full
+mathematical explanation.
+```
+
+```{admonition} Note 2
+Check out the numpy reference
+[np.linalg](https://numpy.org/doc/stable/reference/routines.linalg.html) to
+learn more about the linear algebra functions available
+```
+
+```{code-cell} ipython3
+x = np.linalg.solve(A, b)
+print(x)
+```
+
```

Comment:
Again, it would be nice to add some commentary on what the user sees here. What does `x` represent? What values are we supposed to expect?